src/upgrades/mod.rs

```diff
@@ -129,94 +260,175 @@ pub trait UpgradeStore {
 /// and operators can make changes.
-pub async fn post_start_upgrade(config: &Config, server: &KrillServer) -> Result<(), UpgradeError> {
-    if needs_upgrade(&config.data_dir, "cas", KrillVersion::candidate(0, 9, 3, 2)) {
+pub async fn post_start_upgrade(upgrade_versions: &UpgradeVersions, server: &KrillServer) -> Result<(), UpgradeError> {
+    if upgrade_versions.from() < &KrillVersion::candidate(0, 9, 3, 2) {
         info!("Reissue ROAs on upgrade to force short EE certificate subjects in the objects");
         server.force_renew_roas().await.map_err(|e| e.into())
     } else {
         Ok(())
     }
 }
 
-pub async fn update_storage_version(work_dir: &Path) -> Result<(), UpgradeError> {
-    let current = KrillVersion::current();
-
-    if needs_v0_9_0_upgrade(work_dir, "cas") {
-        debug!("Updating version file for cas");
-        file::save_json(&current, &work_dir.join("cas/version"))?;
-    }
-
-    if needs_v0_9_0_upgrade(work_dir, "pubd") {
-        debug!("Updating version file for pubd");
-        file::save_json(&current, &work_dir.join("pubd/version"))?;
-    }
-
-    Ok(())
-}
-
-fn upgrade_data_to_0_9_0(config: Arc<Config>) -> Result<(), UpgradeError> {
-    let work_dir = &config.data_dir;
-    if needs_v0_9_0_upgrade(work_dir, "pubd") {
-        PubdObjectsMigration::migrate(config.clone())?;
-    }
-
-    if needs_v0_9_0_upgrade(work_dir, "cas") {
-        let signer = Arc::new(KrillSigner::build(work_dir)?);
-        let repo_manager = RepositoryManager::build(config.clone(), signer)?;
-
-        CaObjectsMigration::migrate(config, repo_manager)?;
+/// Returns the KrillUpgradeVersion by comparing the versions of the data used by
+/// the "cas" and "pubd" of the current Krill version to the code version. In the
+/// unlikely event that the "cas" and "pubd" stores are in disagreement, then the
+/// highest of the two version is used as the 'current' version. This can only happen
+/// in practice in case one of the two did not have their version updated in the past,
+/// as there can be only one version running.
+fn upgrade_versions(config: &Config) -> Option<UpgradeVersions> {
+    let cas_version = upgrade_versions_ns(&config.data_dir, CASERVER_DIR);
+    let pubd_version = upgrade_versions_ns(&config.data_dir, PUBSERVER_DIR);
+
+    match (cas_version, pubd_version) {
+        (None, None) => None,
+        (Some(upgrade), None) | (None, Some(upgrade)) => Some(upgrade),
+        (Some(upgrade_cas), Some(upgrade_pubd)) => {
+            if upgrade_cas.from() >= upgrade_pubd.from() {
+                Some(upgrade_cas)
+            } else {
+                Some(upgrade_pubd)
+            }
+        }
     }
-
-    Ok(())
 }
 
-fn needs_v0_9_0_upgrade(work_dir: &Path, ns: &str) -> bool {
-    needs_upgrade(work_dir, ns, KrillVersion::release(0, 9, 0))
-}
-
-fn needs_upgrade(work_dir: &Path, ns: &str, before: KrillVersion) -> bool {
+fn upgrade_versions_ns(work_dir: &Path, ns: &str) -> Option<UpgradeVersions> {
```

Comment:
What is "ns" ?

Reply:
name space - used a lot like this in the code for brevity